File: agents/shared/finit_agent/a2a.py
"""A2A (Agent-to-Agent) protocol server implementation.

Implements JSON-RPC 2.0 over HTTP with agent card discovery and health checks.
"""

from __future__ import annotations

import logging
import time
import uuid
from enum import Enum
from typing import Any, Callable, Awaitable

from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Pydantic models
# ---------------------------------------------------------------------------

class TaskState(str, Enum):
    submitted = "submitted"
    working = "working"
    input_required = "input-required"
    completed = "completed"
    failed = "failed"


class AgentSkill(BaseModel):
    id: str
    name: str
    description: str = ""


class AgentCapabilities(BaseModel):
    streaming: bool = False
    pushNotifications: bool = False


class AgentCard(BaseModel):
    name: str
    description: str
    url: str
    version: str = "1.0.0"
    capabilities: AgentCapabilities = AgentCapabilities()
    skills: list[AgentSkill] = []


class MessagePart(BaseModel):
    type: str = "text"
    text: str = ""


class Message(BaseModel):
    role: str = "user"
    parts: list[MessagePart] = []


class TaskStatus(BaseModel):
    state: TaskState = TaskState.completed
    message: str | None = None


class Artifact(BaseModel):
    parts: list[MessagePart] = []
    metadata: dict[str, Any] = {}


class A2AResult(BaseModel):
    """Result returned by an agent handler."""
    status: TaskStatus = TaskStatus()
    artifacts: list[Artifact] = []


class JsonRpcRequest(BaseModel):
    jsonrpc: str = "2.0"
    method: str
    params: dict[str, Any] = {}
    id: str | int | None = None


class JsonRpcError(BaseModel):
    code: int
    message: str
    data: Any = None


# ---------------------------------------------------------------------------
# Handler type
# ---------------------------------------------------------------------------

TaskHandler = Callable[[str, Message, dict[str, Any]], Awaitable[A2AResult]]


# ---------------------------------------------------------------------------
# JSON-RPC error codes
# ---------------------------------------------------------------------------

PARSE_ERROR = -32700
INVALID_REQUEST = -32600
METHOD_NOT_FOUND = -32601
INTERNAL_ERROR = -32603


def _jsonrpc_error(req_id: str | int | None, code: int, message: str, data: Any = None) -> dict:
    resp: dict[str, Any] = {
        "jsonrpc": "2.0",
        "error": {"code": code, "message": message},
        "id": req_id,
    }
    if data is not None:
        resp["error"]["data"] = data
    return resp


def _jsonrpc_result(req_id: str | int | None, result: Any) -> dict:
    return {
        "jsonrpc": "2.0",
        "result": result,
        "id": req_id,
    }


# ---------------------------------------------------------------------------
# App factory
# ---------------------------------------------------------------------------

_start_time: float = 0.0


def create_a2a_app(agent_card: AgentCard, handler: TaskHandler) -> FastAPI:
    """Create a FastAPI application implementing the A2A protocol.

    Args:
        agent_card: The agent card served at /.well-known/agent.json.
        handler: Async function called for tasks/send.
                 Signature: async def handler(task_id, message, metadata) -> A2AResult
    """
    app = FastAPI(title=f"Finit Agent - {agent_card.name}", version=agent_card.version)
    card_dict = agent_card.model_dump()

    global _start_time
    _start_time = time.time()

    # -- Agent card endpoint --------------------------------------------------

    @app.get("/.well-known/agent.json")
    async def get_agent_card() -> dict:
        return card_dict

    # -- Health check ---------------------------------------------------------

    @app.get("/health")
    async def health() -> dict:
        uptime = time.time() - _start_time
        return {
            "status": "healthy",
            "agent": agent_card.name,
            "uptime_seconds": round(uptime, 2),
        }

    # -- JSON-RPC 2.0 dispatcher ----------------------------------------------

    @app.post("/")
    async def jsonrpc_dispatch(request: Request) -> JSONResponse:
        try:
            body = await request.json()
        except Exception:
            return JSONResponse(
                _jsonrpc_error(None, PARSE_ERROR, "Parse error"),
                status_code=200,
            )

        # Validate basic JSON-RPC structure
        if not isinstance(body, dict) or body.get("jsonrpc") != "2.0" or "method" not in body:
            return JSONResponse(
                _jsonrpc_error(body.get("id") if isinstance(body, dict) else None,
                               INVALID_REQUEST, "Invalid request"),
                status_code=200,
            )

        req_id = body.get("id")
        method = body["method"]
        params = body.get("params", {})

        if method == "tasks/send":
            return await _handle_tasks_send(req_id, params, handler)
        elif method == "tasks/get":
            return await _handle_tasks_get(req_id, params)
        elif method == "tasks/cancel":
            return await _handle_tasks_cancel(req_id, params)
        else:
            return JSONResponse(
                _jsonrpc_error(req_id, METHOD_NOT_FOUND, f"Method not found: {method}"),
                status_code=200,
            )

    return app


async def _handle_tasks_send(
    req_id: str | int | None,
    params: dict[str, Any],
    handler: TaskHandler,
) -> JSONResponse:
    """Handle tasks/send: delegate to the agent handler and return result."""
    task_id = params.get("id", str(uuid.uuid4()))
    raw_message = params.get("message", {})
    metadata = params.get("metadata", {})

    # Parse the message
    try:
        message = Message(**raw_message)
    except Exception:
        message = Message(role="user", parts=[MessagePart(type="text", text=str(raw_message))])

    try:
        logger.info("tasks/send task_id=%s", task_id)
        result = await handler(task_id, message, metadata)

        return JSONResponse(
            _jsonrpc_result(req_id, {
                "id": task_id,
                "status": result.status.model_dump(),
                "artifacts": [a.model_dump() for a in result.artifacts],
            }),
            status_code=200,
        )
    except Exception as exc:
        logger.exception("Handler error for task %s", task_id)
        return JSONResponse(
            _jsonrpc_result(req_id, {
                "id": task_id,
                "status": {"state": "failed", "message": str(exc)},
                "artifacts": [],
            }),
            status_code=200,
        )


async def _handle_tasks_get(req_id: str | int | None, params: dict[str, Any]) -> JSONResponse:
    """Handle tasks/get (stub for PoC)."""
    task_id = params.get("id", "unknown")
    return JSONResponse(
        _jsonrpc_result(req_id, {
            "id": task_id,
            "status": {"state": "completed"},
            "artifacts": [],
        }),
        status_code=200,
    )


async def _handle_tasks_cancel(req_id: str | int | None, params: dict[str, Any]) -> JSONResponse:
    """Handle tasks/cancel (stub for PoC)."""
    task_id = params.get("id", "unknown")
    return JSONResponse(
        _jsonrpc_result(req_id, {
            "id": task_id,
            "status": {"state": "failed", "message": "cancelled"},
            "artifacts": [],
        }),
        status_code=200,
    )

File: orchestrator/src/a2a.rs
use anyhow::{anyhow, Context, Result};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::time::Duration;
use uuid::Uuid;

/// A2A JSON-RPC 2.0 client for communicating with agents.
#[derive(Clone)]
pub struct A2AClient {
    http: Client,
}

// ─── JSON-RPC 2.0 types ─────────────────────────────────────────────────────

#[derive(Debug, Serialize)]
pub struct JsonRpcRequest {
    pub jsonrpc: String,
    pub method: String,
    pub params: TaskSendParams,
    pub id: String,
}

#[derive(Debug, Serialize)]
pub struct TaskSendParams {
    pub id: String,
    pub message: A2AMessage,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct A2AMessage {
    pub role: String,
    pub parts: Vec<A2APart>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum A2APart {
    #[serde(rename = "text")]
    Text { text: String },
}

#[derive(Debug, Deserialize)]
pub struct JsonRpcResponse {
    pub jsonrpc: String,
    pub result: Option<A2ATaskResult>,
    pub error: Option<JsonRpcError>,
    pub id: String,
}

#[derive(Debug, Deserialize)]
pub struct JsonRpcError {
    pub code: i64,
    pub message: String,
    pub data: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct A2ATaskResult {
    pub id: String,
    pub status: A2ATaskStatus,
    #[serde(default)]
    pub artifacts: Vec<A2AArtifact>,
    #[serde(default)]
    pub metadata: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct A2ATaskStatus {
    pub state: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<A2AMessage>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct A2AArtifact {
    #[serde(default)]
    pub parts: Vec<A2APart>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

/// Agent card as returned by /.well-known/agent.json
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentCard {
    pub name: String,
    #[serde(default)]
    pub description: String,
    pub url: String,
    #[serde(default)]
    pub version: String,
    #[serde(default)]
    pub capabilities: Value,
    #[serde(default)]
    pub skills: Vec<AgentSkill>,
    #[serde(default, rename = "securitySchemes")]
    pub security_schemes: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentSkill {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub description: String,
}

impl A2AClient {
    pub fn new(timeout: Duration) -> Self {
        let http = Client::builder()
            .timeout(timeout)
            .build()
            .expect("failed to build HTTP client");
        Self { http }
    }

    /// Discover an agent by fetching its Agent Card from /.well-known/agent.json.
    pub async fn discover(&self, agent_url: &str) -> Result<AgentCard> {
        let url = format!("{}/.well-known/agent.json", agent_url.trim_end_matches('/'));
        let resp = self
            .http
            .get(&url)
            .send()
            .await
            .with_context(|| format!("failed to fetch agent card from {}", url))?;

        if !resp.status().is_success() {
            return Err(anyhow!(
                "agent card request failed: {} {}",
                resp.status(),
                resp.text().await.unwrap_or_default()
            ));
        }

        let card: AgentCard = resp
            .json()
            .await
            .with_context(|| "failed to parse agent card JSON")?;

        Ok(card)
    }

    /// Send a task to an agent using A2A tasks/send.
    pub async fn send_task(
        &self,
        agent_url: &str,
        jwt_token: &str,
        task_id: &str,
        message_text: &str,
        metadata: Option<Value>,
    ) -> Result<A2ATaskResult> {
        let request = JsonRpcRequest {
            jsonrpc: "2.0".to_string(),
            method: "tasks/send".to_string(),
            params: TaskSendParams {
                id: task_id.to_string(),
                message: A2AMessage {
                    role: "user".to_string(),
                    parts: vec![A2APart::Text {
                        text: message_text.to_string(),
                    }],
                },
                metadata,
            },
            id: format!("req-{}", Uuid::new_v4()),
        };

        let resp = self
            .http
            .post(agent_url)
            .header("Authorization", format!("Bearer {}", jwt_token))
            .header("Content-Type", "application/json")
            .json(&request)
            .send()
            .await
            .with_context(|| format!("failed to send task to agent at {}", agent_url))?;

        if !resp.status().is_success() {
            let status = resp.status();
            let body = resp.text().await.unwrap_or_default();
            return Err(anyhow!(
                "A2A request failed: {} {}",
                status,
                body
            ));
        }

        let rpc_response: JsonRpcResponse = resp
            .json()
            .await
            .with_context(|| "failed to parse A2A JSON-RPC response")?;

        if let Some(err) = rpc_response.error {
            return Err(anyhow!(
                "A2A JSON-RPC error {}: {}",
                err.code,
                err.message
            ));
        }

        rpc_response
            .result
            .ok_or_else(|| anyhow!("A2A response missing result"))
    }

    /// Check agent health by calling GET {url}/health.
    pub async fn check_health(&self, agent_url: &str, timeout: Duration) -> Result<bool> {
        let url = format!("{}/health", agent_url.trim_end_matches('/'));
        let client = Client::builder()
            .timeout(timeout)
            .build()
            .expect("failed to build health check client");

        match client.get(&url).send().await {
            Ok(resp) => Ok(resp.status().is_success()),
            Err(_) => Ok(false),
        }
    }
}

/// Extract text from A2A artifacts.
pub fn extract_artifact_text(artifacts: &[A2AArtifact]) -> Option<String> {
    for artifact in artifacts {
        for part in &artifact.parts {
            match part {
                A2APart::Text { text } => return Some(text.clone()),
            }
        }
    }
    None
}

/// Extract text from an A2A task result (from artifacts or status message).
pub fn extract_result_text(result: &A2ATaskResult) -> Option<String> {
    // Try artifacts first
    if let Some(text) = extract_artifact_text(&result.artifacts) {
        return Some(text);
    }
    // Fall back to status message
    if let Some(msg) = &result.status.message {
        for part in &msg.parts {
            match part {
                A2APart::Text { text } => return Some(text.clone()),
            }
        }
    }
    None
}
